src/messenger/messenger.cpp

```diff
@@ -0,0 +1,256 @@
+  routing_id_ = ZmqUtil::GetRoutingId(common::ManagedPointer(socket_));
+
+  socket_->connect(target.GetDestination());
+}
+
+ConnectionId::~ConnectionId() = default;
+
+Messenger::Messenger(common::ManagedPointer<MessengerLogic> messenger_logic) : messenger_logic_(messenger_logic) {
+  // Create a ZMQ context. A ZMQ context abstracts away all of the in-process and networked sockets that ZMQ uses.
+  // The ZMQ context is also the transport for in-process ("inproc") sockets.
+  // Generally speaking, a single process should only have a single ZMQ context.
+  // To have two ZMQ contexts is to have two separate ZMQ instances running, which is unlikely to be desired behavior.
+  zmq_ctx_ = std::make_unique<zmq::context_t>();
+
+  // Register a ROUTER socket on the default Messenger port.
+  // A ROUTER socket is an async server process.
+  zmq_default_socket_ = std::make_unique<zmq::socket_t>(*zmq_ctx_, ZMQ_ROUTER);
+  // By default, the ROUTER socket silently discards messages that cannot be routed.
+  // By setting ZMQ_ROUTER_MANDATORY, the ROUTER socket errors with EHOSTUNREACH instead.
+  zmq_default_socket_->setsockopt(ZMQ_ROUTER_MANDATORY, 1);
+
+  // Bind the same ZeroMQ socket over the default TCP, IPC, and in-process channels.
+  {
+    zmq_default_socket_->bind(MESSENGER_DEFAULT_TCP);
+    zmq_default_socket_->bind(MESSENGER_DEFAULT_IPC);
+    zmq_default_socket_->bind(MESSENGER_DEFAULT_INPROC);
+  }
+
+  // The following block of code is dead code that contains useful background information on ZMQ defaults.
+  {
+    // ZeroMQ does all I/O in background threads. By default, a ZMQ context has one I/O thread.
+    // An old version of the ZMQ guide suggests one I/O thread per GBps of data.
+    // If I/O bottlenecks due to huge message volume, adjust this.
+    // zmq_ctx_set(zmq_ctx_, ZMQ_IO_THREADS, NUM_IO_THREADS);
+  }
+
+  messenger_running_ = true;
+}
+
+Messenger::~Messenger() = default;
+
+void Messenger::RunTask() {
+  // Run the server loop.
+  ServerLoop();
+}
+
+void Messenger::Terminate() {
+  messenger_running_ = false;
+  // TODO(WAN): zmq cleanup? or is that all handled with RAII?
+}
+
+void Messenger::ListenForConnection(const ConnectionDestination &target) {
+  zmq_default_socket_->bind(target.GetDestination());
+}
+
+ConnectionId Messenger::MakeConnection(const ConnectionDestination &target, std::optional<std::string> identity) {
+  const auto &identifier = identity.has_value() ? identity.value() : fmt::format("conn{}", connection_id_count_++);
+  return ConnectionId(common::ManagedPointer(zmq_ctx_), target, identifier);
+}
+
+void Messenger::SendMessage(common::ManagedPointer<ConnectionId> connection_id, std::string message) {
+  ZmqMessage msg{connection_id->routing_id_, message};
+  ZmqUtil::SendMsg(common::ManagedPointer(connection_id->socket_), msg);
+}
+
+void Messenger::ServerLoop() {
+  common::ManagedPointer<zmq::socket_t> socket{zmq_default_socket_};
```

Comment:
If it's the `zmq_default_socket_` that the server loop will always be recving from. How about those sockets opened with `MakeConnection`?

Reply:
ahhh. My bad. I should be doing a poll. Addressed in latest push.